lib/kafka/connection.rb

```diff
@@ -172,23 +230,30 @@ def read_response(response_class, notification)
 
       return correlation_id, response
     rescue Errno::ETIMEDOUT
-      @logger.error "Timed out while waiting for response #{@correlation_id}"
+      @logger.error "Timed out while waiting for response #{expected_correlation_id}"
       raise
     end
 
-    def wait_for_response(response_class, notification)
+    def wait_for_response(response_class, expected_correlation_id, notification)
       loop do
-        correlation_id, response = read_response(response_class, notification)
-
-        # There may have been a previous request that timed out before the client
-        # was able to read the response. In that case, the response will still be
-        # sitting in the socket waiting to be read. If the response we just read
-        # was to a previous request, we can safely skip it.
-        if correlation_id < @correlation_id
-          @logger.error "Received out-of-order response id #{correlation_id}, was expecting #{@correlation_id}"
-        elsif correlation_id > @correlation_id
-          raise Kafka::Error, "Correlation id mismatch: expected #{@correlation_id} but got #{correlation_id}"
+        correlation_id, response = read_response(response_class, expected_correlation_id, notification)
+
+        if correlation_id < expected_correlation_id
+          # There may have been a previous request that timed out before the client
+          # was able to read the response. In that case, the response will still be
+          # sitting in the socket waiting to be read. If the response we just read
+          # was to a previous request, we deliver it to the pending async response
+          # future.
+          async_response = @pending_async_responses.delete(correlation_id)
+          async_response.deliver(response) if async_response
```

Comment:
Hmm... I think it could actually be possible to use `IO.select` to pick the first future that's available:

```ruby
class IoFuture
  def initialize(io, &block)
    @io, @block = io, block
  end

  def to_io # IO.select calls this
    @io
  end

  def value
    @value ||= @block.call
  end
end

futures = [IoFuture.new(socket1) { ... }, IoFuture.new(socket2) { ... }]

until futures.empty?
  ready, _, _ = IO.select(futures, nil, nil, timeout)
  ready.each do |future|
    futures.delete(future)
    response = future.value
    ...
  end
end
```